sympy/polys/matrices/sdm.py

```diff
@@ -28,6 +28,24 @@ def __init__(self, elemsdict, shape, domain):
         if not all(0 <= c < n for row in self.values() for c in row):
             raise DDMBadInputError("Column out of range")
 
+    def getitem(self, i, j):
+        try:
+            return self[i][j]
+        except KeyError:
+            return self.domain.zero
+
+    def extract_slice(self, slice1, slice2):
+        m, n = self.shape
+        ri = range(m)[slice1]
+        ci = range(n)[slice2]
+
+        sdm = {}
+        for i, row in self.items():
+            if i in ri:
+                sdm[ri.index(i)] = {ci.index(j): e for j, e in row.items() if j in ci}
```

Comment:
This should check in case the dict is empty. An SDM should only store non-empty dicts.

Reply:
I am not sure if we require it as it treats empty `dict` as zero in it. Thus every time we pass an empty `dict`, it will treat it as `zero` and make the `DomainMatrix` accordingly.
```
In [2]: DomainMatrix({0: {}, 1: {0: 1}}, (2,1), ZZ)
Out[2]: DomainMatrix([[0], [1]], (2, 1), ZZ)

In [3]: DomainMatrix({0: {0: 0}, 1: {0: 1}}, (2,1), ZZ)
Out[3]: DomainMatrix([[0], [1]], (2, 1), ZZ)

In [4]: DomainMatrix({0: {}, 1: {0: 1}}, (2,2), ZZ)
Out[4]: DomainMatrix([[0, 0], [1, 0]], (2, 2), ZZ)
```
Kindly share your opinion.
Thanks.

Reply:
The empty dict should not be there as it messes up the internal invariant of `DomainMatrix`.
```python
In [7]: M1 = DomainMatrix({0: {}, 1: {0: 1}}, (2,1), ZZ)

In [8]: M2 = DomainMatrix({1: {0: 1}}, (2,1), ZZ)

In [9]: M1
Out[9]: DomainMatrix([[0], [1]], (2, 1), ZZ)

In [10]: M2
Out[10]: DomainMatrix([[0], [1]], (2, 1), ZZ)

In [11]: M1 == M2
Out[11]: False
```

Reply:
Sorry, I assumed they would be equal. Thanks for pointing it out. 
Will fix it.
Thanks.

Reply:
@oscarbenjamin, we have fixed the issue for the `extract_slice`, but still, `SDM` would present `empty` as `zero`:
```
In [1]: from sympy.polys.matrices.domainmatrix import DomainMatrix, DomainScalar, DDM,SDM

In [2]: M1 = DomainMatrix({0: {}, 1: {}, 2: {0: 1}}, (3,1), ZZ)

In [3]: M1[0:, 0]
Out[3]: DomainMatrix([[1]], (1, 1), ZZ)

In [4]: M1
Out[4]: DomainMatrix([[0], [0], [1]], (3, 1), ZZ)
```
Should we open another issue for this problem? 
I mean to say `M1 `should have been just `DomainMatrix([[1]], (1, 1), ZZ)` instead of `DomainMatrix([[0], [0], [1]], (3, 1), ZZ)`.

Kindly share your opinion.
Thanks.